zerver/views/upload.py

```diff
@@ -225,9 +241,82 @@ def serve_image_error(status: int, image_path: str) -> HttpResponseBase:
         patch_vary_headers(response, ("Accept",))
         return response
     if url_only:
+        # TODO: Does this need to handle thumbnails?
         url = generate_unauthed_file_access_url(path_id)
         return json_success(request, data=dict(url=url))
 
+    if thumbnail_format is not None:
+        # Check if this is something that we thumbnail at all
+        try:
+            image_attachment = ImageAttachment.objects.get(path_id=path_id)
+        except ImageAttachment.DoesNotExist:
+            return serve_image_error(404, "images/errors/image-not-exist.png")
+
+        # Validate that this is a potential thumbnail format
+        requested_format = BaseThumbnailFormat.from_string(thumbnail_format)
+        if requested_format is None:
+            return serve_image_error(404, "images/errors/image-not-exist.png")
+
+        rendered_formats = [StoredThumbnailFormat(**f) for f in image_attachment.thumbnail_metadata]
+
+        # We never generate animated versions if the input was still,
+        # so filter those out
+        if image_attachment.frames == 1:
+            potential_output_formats = [
+                thumbnail_format
+                for thumbnail_format in THUMBNAIL_OUTPUT_FORMATS
+                if not thumbnail_format.animated
+            ]
+        else:
+            potential_output_formats = THUMBNAIL_OUTPUT_FORMATS
+        if requested_format not in potential_output_formats:
+            if rendered_formats == []:
+                # We haven't rendered anything, and they requested
+                # something we don't support.
+                return serve_image_error(404, "images/errors/image-not-exist.png")
+
+            accepted_types = sorted(
+                request.accepted_types,
+                key=lambda e: float(e.params.get("q", "1.0")),
+                reverse=True,
+            )
+
+            def q_for(content_type: str) -> float:
+                for potential_type in accepted_types:
+                    if potential_type.match(content_type):
+                        return float(potential_type.params.get("q", "1.0"))
+                return 0.0
+
+            # Serve a "close" format -- preferring animated which
+            # matches, followed by the format they requested, or one
+            # their browser supports, in the size closest to what they
+            # requested, with the minimum bytes.
+            def grade_format(
+                possible_format: StoredThumbnailFormat,
+            ) -> Tuple[bool, bool, float, int, int]:
+                assert requested_format is not None
+                return (
+                    possible_format.animated != requested_format.animated,
+                    possible_format.extension != requested_format.extension,
+                    1.0 - q_for(possible_format.content_type),
+                    abs(requested_format.max_width - possible_format.max_width),
+                    possible_format.byte_size,
+                )
```

Comment:
Probably worth making `grade_format` a top-level function so we can have a bit of a test suite for it, since it's fairly dense.